src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MSITest.java

```diff
@@ -0,0 +1,189 @@
+        // unregister the custom providers registered in AESetup
+        SQLServerConnection.unregisterColumnEncryptionKeyStoreProviders();
+
+        // set to use Managed Identity for keystore auth
+        String connStr = AETestConnectionString;
+        connStr = TestUtils.addOrOverrideProperty(connStr, Constants.KEYSTORE_AUTHENTICATION,
+                "KeyVaultManagedIdentity");
+        testNumericAKV(connStr);
+    }
+
+    private void testCharAkv(String connStr) throws SQLException {
+        String sql = "select * from " + CHAR_TABLE_AE;
+        try (SQLServerConnection con = PrepUtil.getConnection(connStr);
+                SQLServerStatement stmt = (SQLServerStatement) con.createStatement();
+                SQLServerPreparedStatement pstmt = (SQLServerPreparedStatement) TestUtils.getPreparedStmt(con, sql,
+                        stmtColEncSetting)) {
+            TestUtils.dropTableIfExists(CHAR_TABLE_AE, stmt);
+            createTable(CHAR_TABLE_AE, cekAkv, charTable);
+            String[] values = createCharValues(false);
+            populateCharNormalCase(values);
+
+            try (ResultSet rs = (null == stmt) ? pstmt.executeQuery() : stmt.executeQuery(sql)) {
+                int numberOfColumns = rs.getMetaData().getColumnCount();
+                while (rs.next()) {
+                    AECommon.testGetString(rs, numberOfColumns, values);
+                    AECommon.testGetObject(rs, numberOfColumns, values);
+                }
+            }
+        }
+    }
+
+    private void testNumericAKV(String connStr) throws SQLException {
+        String sql = "select * from " + NUMERIC_TABLE_AE;
+        try (SQLServerConnection con = PrepUtil.getConnection(connStr);
+                SQLServerStatement stmt = (SQLServerStatement) con.createStatement();
+                SQLServerPreparedStatement pstmt = (SQLServerPreparedStatement) TestUtils.getPreparedStmt(con, sql,
+                        stmtColEncSetting)) {
+            TestUtils.dropTableIfExists(NUMERIC_TABLE_AE, stmt);
+            createTable(NUMERIC_TABLE_AE, cekAkv, numericTable);
+            String[] values = createNumericValues(false);
+            populateNumeric(values);
+
+            try (SQLServerResultSet rs = (null == stmt) ? (SQLServerResultSet) pstmt.executeQuery()
+                                                        : (SQLServerResultSet) stmt.executeQuery(sql)) {
+                int numberOfColumns = rs.getMetaData().getColumnCount();
+                while (rs.next()) {
+                    AECommon.testGetString(rs, numberOfColumns, values);
+                    AECommon.testGetObject(rs, numberOfColumns, values);
+                    AECommon.testGetBigDecimal(rs, numberOfColumns, values);
+                    AECommon.testWithSpecifiedtype(rs, numberOfColumns, values);
+                }
+            } catch (Exception e) {
+                fail(TestResource.getResource("R_loginFailed") + e.getMessage());

```

Comment:
Login failed? This catch block implies that, the only exception that can happen inside try block is due to a login failure. Just remove the catch block and let the test throw the original exception.